### DISCUSSION TITLE
_tkinter.TclError when using Python 3.12 and 3.13 from Homebrew on macOS Sequoia 15.2

### DISCUSSION THREAD

**Discussion by w8sl** [Homebrew/discussions#5809] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: >=4.3.0 (shallow or no git repository)
ORIGIN: (none)
HEAD: (none)
Last commit: never
Branch: (none)
Core tap HEAD: (none)
Core tap last commit: never
Core tap branch: (none)
Core tap JSON: 17 Dec 18:43 UTC
Core cask tap JSON: 17 Dec 18:43 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.6 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.2-arm64
CLT: 16.2.0.0.1.1733547573
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew
```

### Description of issue

I get: _tkinter.TclError: bad option "variable": must be add, info, or remove, which is related to deprecated function: self.trace("w",..)
Correcting all occurrences in the source code of https://github.com/oscarpilote/Ortho4XP to: self.trace_add("write",..)  doesn´t help! Still impossible to make previews and move the map. Unfortunately, it is very difficult to reproduce the issue with simple code example.
Interestingly, everything (as far as I can see) is ok on macOS Sequoia, with with Python 3.13 and 3.14 from python.org (self.trace("w",..) is accepted btw)

**Comment by Bo98** [comment]
`trace_add` should work. `trace`/`trace_variable` indeed has been deprecated since 2016 (Python 3.6).

**Reply by w8sl** [reply]
I suspect trace is not the only problem. Something weird is happening after correction, like ImageTk.PhotoImage cannot find the image when called from a complex function like this:  
https://github.com/oscarpilote/Ortho4XP/blob/c5e0d4b50774de1850246fac56b9d461ae026a72/src/O4_GUI_Utils.py#L894
Simpler code works without problems
This is very specific for Homebrews Python 3.12 and 3.13 for Sequoia. Everything else works on Windows,Linux and Python from org on Sequoia.

**Comment by w8sl** [comment]
There’s a threading issue also. Here’s an example code that reproduces the crash (_tkinter.TclError: image type "photo" does not exist) with Homebrew’s Python 3.13 on Sequoia 15.2, but works with any Python 3.11 or 3.13 from  python.org:

(requires pillow)
<pre>

import os
import tkinter as tk
from PIL import Image, ImageTk
import traceback
import threading
import queue

class Example:
    def __init__(self, root):
        self.root = root
        self.photo = None
        self.queue = queue.Queue()
        self.image_label = tk.Label(root)
        self.image_label.pack()
        self.check_queue()
        # Start the image loading process in a separate thread
        threading.Thread(target=self.async_build_map_layer, args=(37, 23, 11, "OSM", self.queue)).start()
    
    def check_queue(self):
        try:
            photo = self.queue.get_nowait()
            self.update_image(photo)
        except queue.Empty:
            self.root.after(100, self.check_queue)
    
    def update_image(self, photo):
        self.photo = photo
        self.image_label.config(image=self.photo)
        self.image_label.image = self.photo  # Keep a reference to avoid garbage collection

    @staticmethod
    def async_build_map_layer(lat, lon, zl, provider, output_queue):
        try:
            # Simulate loading an image file
            image_path = "absolute path to some .jpeg"
            background_map = Image.open(image_path)
            print("Image loaded successfully")
            print(lat, lon, zl, provider)

            # Simulate some processing delay
            import time
            time.sleep(2)
            
            photo = ImageTk.PhotoImage(background_map)
            output_queue.put(photo)
        
        except Exception as e:
            traceback.print_exc()
            raise e

root = tk.Tk()
app = Example(root)
root.mainloop()

**Reply by cho-m** [reply, marked as answer]
If you are able to, try building https://github.com/Homebrew/homebrew-core/pull/201830 from source and see if it works.

That applies part of upstream commit https://github.com/python/cpython/commit/47cbf038850852cdcbe7a404ed7c64542340d58a for some more compatibility with TCL 9 along with my open PR to fix threading.

---

If there are still issues, will need to root cause to figure out where they are happening (e.g. in case of pillow, it could be in pillow calls to tkinter API, CPython tkinter's calls to TCL/TK API, or in TCL/TK itself)

**Comment by w8sl** [comment]
That fixes the problem with sample code and Ortho4XP previews as well - thank you !!!

Changes not present in Python 3.13 from org are introduced: 

* trace("w") -> trace_add("write") which is backwards compatible and well described in "Migrating scripts to Tcl 9" documentation

* change of right mouse click behavior from: "ButtonPress-2" and "B2-Motion" on macOS, "ButtonPress-3" and  "B3-Motion" on other systems, into: "ButtonPress-3" and "B3-Motion" on all systems, which is not backwards compatible !

(Python 3.12 is also affected by the same problem, Python 3.11 is ok)

**Reply by w8sl** [reply]
While we can just update code to trace_add("write") and forget, changed mouse behavior is quite a mess. Was it intentional? Python 13.1 x86-64 on GitHub runner is TCL 8!

**Reply by Bo98** [reply]
Python versions make no guarantees on the underlying Tcl/Tk version:

> Tkinter supports a range of Tcl/Tk versions, built either with or without thread support

Various Linux distributions of Python will also migrate to Tcl/Tk 9 on different dates, depending on their own release schedules or support policies.

Given you likely already have macOS conditionals for those buttons, the change here is to modify those conditionals to include `and tkinter.TkVersion < 9`.

It's unfortunate this change has broken your code and I am sorry for that. The change would land at some point however so would have been a matter of delaying the issue - the timing of Tcl 9 migration is unlikely to exactly align with every other build of Python (e.g. Fedora is considering performing the migration with their next release in April). Even GitHub Actions I think has its own builds of Python for some platforms.

**Reply by w8sl** [reply]
>
> Given you likely already have macOS conditionals for those buttons, the change here is to modify those conditionals to include `and tkinter.TkVersion < 9`.
> 

Thank you for very helpful suggestion!

**Reply by spyoungtech** [reply]
As far as I can tell, CPython [does not](https://github.com/python/cpython/issues/124111) yet [support](https://github.com/python/cpython/issues/104568) tcl/tk 9 and has known issues on MacOS (or even newer versions of 8.x beyond 8.6, it seems).

It is true that there is no "guarantee" of a specific tcl/tk version, I am not sure that means that it's reasonable to use any version of tcl/tk with Python (not unlike how using different versions/configurations of many other linked library can break a Python build in other unexpected ways). For example, there are stated [minimum supported versions](https://github.com/python/cpython/blob/c4d37eefb7b33f10f776183447af44faf8efa7f6/Modules/_tkinter.c#L12-L14) -- it should also be plain to understand that the "range of tcl/tk versions" supported can obviously have an upper-bound, too, for any given version of Python... as seems to be the case today with tcl/tk 9 and all versions of Python up to the most recent alpha release of 3.14.

Although 3.14 includes many (but perhaps not all?) changes needed for support of newer tcl/tk versions, the [proposed backports](https://github.com/python/cpython/pull/127364) for 3.12 and 3.13 have been rejected/withdrawn -- so it seems unlikely Python3.12 or Python3.13 will ever be officially supported with tcl/tk 9.

If we look at the rest of the ecosystem supporting python/tkinter, they are all still statically linked to (or their associated tk packages depend on) tcl/tk 8.6.x for these Python versions, even in up to 3.14:

- Official builds all bundle tcl/tk 8.6 for every platform, including Python3.14.0a5 which still uses 8.6
- Ubuntu [`python3.13-tk`](https://launchpad.net/ubuntu/oracular/amd64/python3.13-tk/3.13.0-1) is using 8.6
- Alpine [python3-tkinter](https://pkgs.alpinelinux.org/package/edge/community/x86/python3-tkinter) is using 8.6 (for Python 3.12, no 3.13 package is provided, as far as I know)
- [Fedora's python3.13-tkinter](https://packages.fedoraproject.org/pkgs/python3.13/python3.13-tkinter/fedora-40-updates.html) is using [tcl 8.6.13](https://packages.fedoraproject.org/pkgs/tcl/tcl/fedora-40.html) (moreover, even after providing tcl/tk 9, in recognition of the breaking changes and support needed, fedora [will continue to provide tcl/tk 8 packages](https://fedoraproject.org/wiki/Changes/TclTk9.0#Summary) which we have every reason to suspect will continue to be the dependent version in future python3.12/3.13 packages)
- Arch Linux [python3.13.2](https://archlinux.org/packages/core/x86_64/python/)'s tk optional is using [tk 8.6.16](https://archlinux.org/packages/extra/x86_64/tk/)
- [Deadsnakes](https://github.com/deadsnakes/py3.13) is also still using tcl/tk 8.x in 3.13 and 3.14
- Indygreg/astral/uv (python-standalone-build) builds _just_ [updated to 8.6.14](https://github.com/astral-sh/python-build-standalone/pull/495) earlier this year

Although new versions of tcl/tk may not align directly with python releases, package maintainers and third party Python builds will largely  continue to have the ability to stick with an appropriate versions, as I see it.

The homebrew formula for `python-tk` Python3.12 and 3.13 seems to be the only major outlier here and, in my personal experience maintaining and using packages using tkinter, it seems to uniquely be MacOS brew users who are most commonly affected by such problems. Users using official builds of Python3.12 and 3.13 on MacOS or other builds on other platforms do not seem to have these problems because they are using the same expected tcl/tk version.

**In summary:**

1. CPython 3.12 and 3.13 ostensibly do not fully support tcl/tk 9
2. A backport providing tcl/tk 9 support in these Python versions will likely not happen in the future
3. Official builds bundling tcl/tk and virtually every third-party `python3-tk[inter]` package is still using tcl/tk 8.6.x for Python3.12 and 3.13
4. MacOS homebrew users seem to constitute the bulk of the cohort of users having compatibility issues with tkinter in Python3.12 and 3.13


Given the foregoing, I would ask that you please consider adding the same tcl/tk `@8` constraint for the `python-tk` formula versions 3.12 and 3.13, as exist for previous versions of the formula to maintain harmony with official builds and expectations users have with other `python-tk[inter]` third party packages -- at least for Python 3.12 and 3.13, if not also 3.14.

Thank you for your continued support of these formulas and for your consideration on this issue.
